docs/RELEASE.md

```diff
@@ -0,0 +1,126 @@
+
+## Versioning
+
+We are releasing multiple different things from this repository in one release, but 
+we don't want to use the same version for everything. Thus, in the following we explain
+the versioning story for the crates, node and Westend & Rococo. To easily refer to a release, we shall use the node version of it.
+
+### Crate
+
+We try to follow SemVer<sup>3</sup> as best as possible for versioning our crates' public APIs.  
+
+👉 The public API of our library crates is defined as all public items that are not inside a `__private` module.
+
+### Node
+
+The versioning of the node is done 99% of the time by only incrementing the `minor` version. 
+The `major` version is only bumped for special releases and the `patch` can be used for an 
+out of band release that fixes some critical bug. The node version is not following SemVer. 
+This means that the version doesn't express if there are any breaking changes in the CLI 
+interface or similar. The node version is declared in the `NODE_VERSION` variable in 
+`polkadot/node/primitives/src/lib.rs`.
+
+### Westend & Rococo
+
+For the these networks, we only increment the `spec_version` and sometimes the `transaction_version`. The spec version is also following
+the node version. Its schema is: `M_mmm_ppp` and for example `1_002_000` is the node release `1.2.0`. This versioning has no further meaning, and is only done to map from an on chain `spec_version` easily to the release in this repository. 
+
+## Backports
+
+Backports should most of the time not be required. We should only backport [critical bug fixes](#bug-and-security-fix) and then release the fixed crates. There should be no need to backport anything from a release branch.
+
+# Processes
+
+The following processes are necessary to actualize our releases. Each process has a *Cadence* on which it must execute and an *Responsible* that is responsible for autonomously doing so and reporting back any error in the RelEng<sup>1</sup> channel.
+
+## Crate Bumping
+
+Cadence: (possibly) each Merge Request. Responsible: Developer that opened the MR.
+
+Following SemVer isn't easy, but there exists [a guide](https://doc.rust-lang.org/cargo/reference/semver.html) in the Rust documentation that explains the small details on when to bump what. This process should be augmented with CI checks that utilize [`cargo-semver-checks`](https://github.com/obi1kenobi/cargo-semver-checks) and/or [`cargo-public-api`](https://github.com/Enselic/cargo-public-api). They must also pay attention to downstream dependencies that require a version bump, because they export the changed API.
+
+### Steps
+
+1. Developer opens a Merge Request with changed crates against `master`.
+2. They bump all changed crates according to SemVer.
+3. They bump all crates that export any changed types in their *public API*.
+4. They also bump all crates that inherit logic changes from relying on one of the bumped crates. 
```

Comment:
I was trying to describe the normal workflow of merging a standard Merge request.  

Currently we never touch the `Cargo.toml` version. But in the future we would need to increment it according to SemVer. This can be a `major`, `minor` or `patch` bump - depending on the change.  

Otherwise when would we bump these versions? We dont really "merge" to the release branch, unless its a backport.

Reply:
My understanding was

- merge to `master`: don't need to do bumps on merge. nightly releases will be made nightly bumping the semver with a new `-nightlyYYYY-MM-DD` suffix. I don't see benefit of bumping X.Y.Z between nightly releases, because the code will have a breaking bump (-nightly change) soon anyway. 
- merge to `release`: we need to bump according to X.Y.Z semver, because we need to make new releases for the changed crates.

Reply:
This is how I understood it

```
stable
v1.0.0 -> v1.0.1 -------------------------------> v2.0.0
                                                    ^
                                             CLOBBER|
master                                              |
v2.0.0-nightly2023-01-01 -> ... v2.0.0-nightly2023-04-01 -> v3.0.0-nightly2023-04-02 -> ...
```

Reply:
> But then how do we keep track of these in the meantime?

Keep track of what in the meantime?

> Note that we dont need to bump multiple times on master if there was no release so far.

Yup, it should only create a new nightly build if there have been changes. That should be automated. 

> We would also need to bullet proof automate these version bump then. I would also like to do this with prdoc and so on, but @bkchr seems to be in favour of just bumping them directly.

I assume you mean automate version bumps in stable?

Patch bumps should be easy enough to automate. I plan to add a subcommand to https://github.com/liamaharon/cargo-workspace-version-tools where you specify a crate to patch, and it'll identify all dependants of that crate (direct or via other crates). Once it has all the dependants, it bumps all the Cargo.toml versions. Then in CI, it should detect Cargo.toml version bumps and release those crates (I can look into a script for this if the release team is unable to).

We just need to be very careful to not accidentally make a breaking change in these patches. I'm thinking we use a combination of those two semver check libraries we've discussed to inspect the public api before/after. Maybe a bot can post a diff of the public API in the PR to make it easier to view.

Reply:
> Patch bumps should be easy enough to automate

This isn't the problem. The problem is more to bump versions properly for breaking changes.



> I would also like to do this with prdoc and so on, but @bkchr seems to be in favour of just bumping them directly.

I'm really not happy with either solution. Prdoc will require to copy X crates names and the version bump it requires. While on the other side we may bump a crate version too often in the `Cargo.toml` on master. I have pushed back for the prdoc option, but if you insist I release my "block" on this. I was imagining that we have some extra CI job that informs us if we for example want to bump the major version multiple times on master.


Orrr.... We just scrap versions completely from master and ensure that when we backport stuff to stable that versions are bumped properly. We could use these semver tools to help us on these bumps. I didn't tried them yet, but we can see how good they work and if not improve them on the way. WDYT?

Reply:
Whatever we do, we loose "visibility" any way on master on which code exactly is on crates.io.

Reply:
> Whatever we do, we loose "visibility" any way on master on which code exactly is on crates.io.

CI has to pick a commit to deploy from anyway, it shouldn't be hard for CI every nightly release to bump the `Cargo.toml` versions YYYY-MM-DD, commit it, and tag the commit?

> Orrr.... We just scrap versions completely from master and ensure that when we backport stuff to stable that versions are bumped properly. We could use these semver tools to help us on these bumps. I didn't tried them yet, but we can see how good they work and if not improve them on the way. WDYT?

My understanding from the forum post is this was the plan yes: `master` is never bumped manually, it's only bumped by the CI every nightly release.

For example, if a crate version in `stable` is 1.2.1, `master` would be set to 2.0.0-nightlyYYYY-MM-DD. 

Then every 6th release, we remove the `-nightlyYYYY-MM-DD` from the versions and that becomes our new `stable` branch. 

This way, we don't need to worry at all about breaking changes when we merge to `master`.

A downside of this process is that we blindly push major bumps for every crates every 6th release even if there were no breaking changes. HOWEVER this is still 10x better than what we do now, so I think it makes sense to get this process out now and make the diminishing returns improvements later iteratively, rather than trying to get something perfect all at once.

Reply:
It could work for `master`, and it does sound quite a lot easier for a start. But then we loose the ability to backport features/improvements, since we have no overview when a breaking change merged to master.

Reply:
> Then every 6th release, we remove the `-nightlyYYYY-MM-DD` from the versions and that becomes our new `stable` branch.

So we would still bump only major every 3 months? 



> On the other hand if we put the crate bumps into the PRdoc, then we can easily backport stuff that has only `minor/patch` in it. Would be even easier than bumping the `Cargo.toml` directly, since the Prdoc gives us the semver-diff.

I mean I see the advantage and maybe we should do it and live with copying over the bumps to the prdoc. The problem is still that we could forget to mention things that are require a bump. I would really first like to try out the tooling and then see what we should do. I mean the tooling probably already works for most of the things we are doing as breaking changes, like changing a function name or removing a parameter.

Reply:
> So we would still bump only major every 3 months?

In my proposal, yes.

In my proposal:

- Major bumps on `release` every 3 months when the Clobber happens. SemVer respected.
- Minor bumps on `release` every 2 weeks for any crate touched or any crate depending on such a crate. No backports are allowed to make breaking changes. SemVer respected.
- Patch bumps on `release` for out-of-band security fixes which have been reviewed as non-breaking. SemVer respected.
- Nightly just has the next major version with a `-nightly` suffix. SemVer respected.

Reply:
> I would really first like to try out the tooling and then see what we should do. I mean the tooling probably already works for most of the things we are doing as breaking changes, like changing a function name or removing a parameter.

Okay then lets try the native bumping without Prdoc for now.